### DISCUSSION TITLE
Reviewing CI failures

### DISCUSSION THREAD

**Discussion by dabd** [mitsuhiko/insta#403] [General]
Since the most common use case for snapshot testing is asserting on large values, what techniques do you use to communicate failures in CI ? Given the default setting for INSTA_UPDATE is auto, I suppose you have to replay the tests locally but what if you can’t reproduce it easily?

**Comment by mitsuhiko** [comment]
I never tried this but you could try to upload the `.new` snapshots into github artifacts and then download them. If that sort of workflow works, it could potentially be automated.

**Comment by max-sixty** [comment]
I'm not sure this would work, but writing an annotation or suggestion from GHA would be quite cool.

It would require parsing the insta output, and then writing it out as [an error message annotation](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-error-message). Or writing a [problem matcher](https://github.com/actions/toolkit/blob/main/docs/problem-matchers.md), which would be easier but less featureful.

I don't think annotations support suggestions, but with some GHA permission wrangling, it could instead be written out as a comment containing a suggestion, which then someone could click "Accept" on in GitHub...

**Comment by dabd** [comment]
BTW, I was trying to understand in detail (from the documentation) the behaviour of each INSTA_UPDATE option.

Are the following flowcharts correct?

They highlight that the underlying workflows are not that simple (at least for me), for example Unseen and Always are potentially a bit like power user options, bypassing review in case of mismatches, since they write `.snap` files directly, instead of `.snap.new` files for subsequent review.

![graphviz (1)](https://github.com/mitsuhiko/insta/assets/167463/8b92632e-dfaf-4d7e-9b23-738e44254971)

![graphviz](https://github.com/mitsuhiko/insta/assets/167463/8f786f89-d1f5-40e0-b90c-aafa40ad0dbf)

![graphviz (4)](https://github.com/mitsuhiko/insta/assets/167463/d772d73b-5c93-4dac-b4d3-4b092e11aa80)

![graphviz (3)](https://github.com/mitsuhiko/insta/assets/167463/e3dad6b8-bb66-4de4-b6fc-3df7324992c6)

![graphviz (2)](https://github.com/mitsuhiko/insta/assets/167463/10bd390e-ebca-45b2-9cb6-0ee3f88baf82)

**Reply by max-sixty** [reply]
(not my library but responding regardless)

I think these are correct!

> for example Unseen and Always are potentially a bit like power user options, bypassing review in case of mismatches, since they write `.snap` files directly, instead of `.snap.new` files for subsequent review.

That's correct — that's very much their intention IIUC. Personally, I generally use `--accept`, similar to `always`[^1], and then use git to review changes.

[^1]: these could do with a consolidation between env vars & flags; touched on in https://github.com/mitsuhiko/insta/issues/368
